doc/man/rdma-ndd.8.in
.\" Man page generated from reStructuredText.
.
.TH RDMA-NDD 8 "@BUILD_DATE@" "" "OpenIB Diagnostics"
.SH NAME
RDMA-NDD \- RDMA device Node Description update daemon
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
rdma\-ndd <options>
.SH DESCRIPTION
.sp
rdma\-ndd is a system daemon which watches for rdma device changes and/or
hostname changes and updates the Node Description of the rdma devices based on
those changes.
.SH DETAILS
.sp
Full operation of this daemon requires kernels which support polling of the
procfs hostname file as well as libudev.
.sp
If your system does not support either of these features, the daemon will set
the Node Descriptions at start up and then sleep forever.
.SS Node Description format
.sp
The daemon uses the nd_format configuration option within the ibdiags.conf
file.  %h and %d can be used as wildcards in that string to specify the dynamic
use of <hostname> and <device> respectively.
.sp
NOTE: At startup and on new device detection the Node Description is always
written to ensure the SM and this daemon are in sync.  Subsequent events will
only write the Node Description on a device if it has changed.
.SH OPTIONS
.sp
\fB\-\-retry_timer, \-t\fP
Length of time to sleep when system errors occur when attempting to poll and or read the hostname from the system.
.sp
\fB\-\-retry_count, \-r\fP
Number of times to attempt to retry setting of the node description on failure.  Default 0
.sp
\fB\-\-foreground, \-f\fP
Run in the foreground instead of as a daemon
.sp
\fB\-\-pidfile <pidfile>\fP
specify a pid file (daemon mode only)
.SS Configuration flags
.\" Define the common option -z
.
.sp
\fB\-\-config, \-z  <config_file>\fP Specify alternate config file.
.INDENT 0.0
.INDENT 3.5
Default: @IBDIAG_CONFIG_PATH@/ibdiag.conf
.UNINDENT
.UNINDENT
.SS Debugging flags
.\" Define the common option -h
.
.sp
\fB\-h, \-\-help\fP      show the usage message
.\" Define the common option -V
.
.sp
\fB\-V, \-\-version\fP     show the version info.
.SH FILES
.\" Common text for the config file
.
.SS CONFIG FILE
.sp
@IBDIAG_CONFIG_PATH@/ibdiag.conf
.sp
A global config file is provided to set some of the common options for all
tools.  See supplied config file for details.
.SH AUTHOR
.INDENT 0.0
.TP
.B Ira Weiny
< \fI\%ira.weiny@intel.com\fP >
.UNINDENT
.\" Generated by docutils manpage writer.
.


ChangeLog
commit ae7c98dba44da7761f1c911332f73da47aab7d3b
Author: Tatyana Nikolova <tatyana.e.nikolova@intel.com>
Date:   Wed May 11 14:15:22 2016 -0700

    Updating version to 1.0.4 to be included in OFED-3.18-2
    
    Signed-off-by: Tatyana Nikolova <tatyana.e.nikolova@intel.com>

commit ca1928eacc419fa7215c5297a0081cccd353b602
Author: Mustafa Ismail <mustafa.ismail@intel.com>
Date:   Wed May 11 14:00:52 2016 -0700

    Fix for freeing a nlmsg after sending it
    
    Signed-off-by: Tatyana Nikolova <tatyana.e.nikolova@intel.com>

commit 3119700891b802517577b189193b7e76ee256508
Author: Mustafa Ismail <mustafa.ismail@intel.com>
Date:   Wed May 11 13:57:22 2016 -0700

    Fix for freeing the list of network interfaces after using it
    
    Signed-off-by: Tatyana Nikolova <tatyana.e.nikolova@intel.com>

commit ab8225cd5dfd64eb6cdde69defcf7593fb429396
Author: Tatyana Nikolova <tatyana.e.nikolova@intel.com>
Date:   Wed May 11 13:56:13 2016 -0700

    Update ChangeLog
    
    Signed-off-by: Tatyana Nikolova <tatyana.e.nikolova@intel.com>

commit 03f0bd35c1804c0171cc013e36c1dc94c131dff9
Author: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>
Date:   Wed Oct 14 10:06:04 2015 -0700

    Updating version to 1.0.3 to be included in OFED-3.18-1
    
    Signed-off-by: Tatyana Nikolova <tatyana.e.nikolova@intel.com>

commit 360337da4b10c0a26edcbf6adaf7eab02570b782
Author: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>
Date:   Tue Oct 13 14:36:13 2015 -0700

    Update ChangeLog
    
    Signed-off-by: Tatyana Nikolova <tatyana.e.nikolova@intel.com>

commit a18b1deca6bb6c0d7b2110bb6fdfff4902489c3f
Author: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>
Date:   Tue Oct 13 14:26:58 2015 -0700

    The iwpmd needs to start after the network service
    
    Fix for the port mapper service not starting
    after boot on SLES11 SP4
    
    Signed-off-by: Tatyana Nikolova <tatyana.e.nikolova@intel.com>

commit c8a4ce71741dbaa3407e96ec0b1ca82337a35e3d
Author: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>
Date:   Tue Aug 18 12:55:38 2015 -0700

    Adding ChangeLog
    
    Signed-off-by: Tatyana Nikolova <tatyana.e.nikolova@intel.com>

commit 5f5be6fe1a0723764a1a7df5880d197a4ccd5d47
Author: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>
Date:   Mon Aug 17 12:54:09 2015 -0700

    Adding a sample iwpmd.conf file
    
    Printing the size of the netlink socket receive
    buffer only if debug is enabled
    
    Removing the dash from the verison number
    
    Releasing libiwpm-1.0.3rc1
    
    Signed-off-by: Tatyana Nikolova <tatyana.e.nikolova@intel.com>

commit bb38d6528e11141914baaa711606e28c9e1a31f0
Author: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>
Date:   Fri Aug 14 14:47:08 2015 -0700

    Updating libiwpm version to 1.0.3-rc1
    
    Signed-off-by: Tatyana Nikolova <tatyana.e.nikolova@intel.com>

commit 095bb4bf58518a6b3783662baab4a4d7042427be
Author: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>
Date:   Fri Aug 14 14:27:58 2015 -0700

    Adding iwpmd.conf file, which allows for changing the size
    of the netlink socket receive buffer
    
    Including a fix submitted by Steve Wise
    for Bug 2559 in openfabrics bugzilla, reproduced by Chelsio
    about failing to send a netlink message to iwpmd
    
    Signed-off-by: Tatyana Nikolova <tatyana.e.nikolova@intel.com>

commit 4f34c76c396d7a5d9d33449145000e1ed46fc630
Author: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>
Date:   Fri Aug 14 12:22:11 2015 -0700

    Changing libiwpm version in configure.ac
    to 1.0.2
    
    Signed-off-by: Tatyana Nikolova <tatyana.e.nikolova@intel.com>

commit 1f436ed6271cc5af8456fa0a372b803553d92b10
Author: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>
Date:   Thu Jul 2 14:01:55 2015 -0700

    Updating the README
    
    Signed-off-by: Tatyana Nikolova <tatyana.e.nikolova@intel.com>

commit 713b27328aed37d00badabb9492bfe91a7ec20ee
Author: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>
Date:   Thu Jul 2 13:16:57 2015 -0700

    Changes to allow multi devices per netlink client
    to register with the port mapper
    
    Signed-off-by: Tatyana Nikolova <tatyana.e.nikolova@intel.com>

commit 672d76b13639da5a35fad6be2c578dc78623605f
Author: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>
Date:   Thu Jul 2 13:06:51 2015 -0700

    Updating libiwpm version to 1.0.2
    for OFED-3.18 release
    
    Signed-off-by: Tatyana Nikolova <tatyana.e.nikolova@intel.com>

commit c42cad0c4b6f84cdb873f093f941dac785a07de1
Author: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>
Date:   Wed May 6 13:30:32 2015 -0700

    From: Steve Wise <swise@opengridcomputing.com>
    
    Fix for make dist to include systemd service file
    
    Signed-off-by: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>

commit b9d08c0e595ba4b1d3e1db40dd26448386fc571e
Author: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>
Date:   Fri May 1 14:49:53 2015 -0700

    Changes to the port mapper spec and start-up script files
    1) Fixes to enable the iwpmd install and start when systemd is available
    2) Adding "ulimit -n 102400" to the start up scripts to allow testing
    with high number of connections
    
    Signed-off-by: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>

commit 7bdb4da6ef40504a36545d2d116ef43a4d694e38
Merge: 21ef408 ac69af0
Author: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>
Date:   Fri May 1 14:44:36 2015 -0700

    Merge branch 'for-next'

commit ac69af064d32b222afd1e155a1d2ccc5305a12aa
Author: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>
Date:   Fri May 1 14:26:22 2015 -0700

    Replacing INCLUDES with AM_CPPFLAGS in the makefile
    Removing autoscan.log
    
    Signed-off-by: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>

commit 21ef408eb146d2eff8016d8a7c1410350254a55e
Author: Tatyana E. Nikolova <tatyana.e.nikolova@intel.com>
Date:   Thu Dec 11 13:52:29 2014 -0800

    New iwpmd service file for systemd on RHEL 7


src/rdma-ndd.c
/*
 * Copyright (c) 2014 Intel Corporation. All Rights Reserved
 *
 * This software is available to you under a choice of one of two
 * licenses.  You may choose to be licensed under the terms of the GNU
 * General Public License (GPL) Version 2, available from the file
 * COPYING in the main directory of this source tree, or the
 * OpenIB.org BSD license below:
 *
 *     Redistribution and use in source and binary forms, with or
 *     without modification, are permitted provided that the following
 *     conditions are met:
 *
 *      - Redistributions of source code must retain the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer.
 *
 *      - Redistributions in binary form must reproduce the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer in the documentation and/or other materials
 *        provided with the distribution.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

#if HAVE_CONFIG_H
#  include <config.h>
#endif				/* HAVE_CONFIG_H */

#include <poll.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <assert.h>
#include <string.h>
#include <limits.h>
#include <stdio.h>
#include <syslog.h>
#include <dirent.h>
#include <errno.h>
#include <unistd.h>
#include <getopt.h>
#include <stdlib.h>

#include <libudev.h>

struct udev *udev;
struct udev_monitor *mon;

#include "ibdiag_common.h"

#define SYS_HOSTNAME "/proc/sys/kernel/hostname"
#define SYS_INFINIBAND "/sys/class/infiniband"
#define DEFAULT_RETRY_RATE 60
#define DEFAULT_RETRY_COUNT 0
#define DEFAULT_ND_FORMAT "%h %d"

int failure_retry_rate = DEFAULT_RETRY_RATE;
int set_retry_cnt = DEFAULT_RETRY_COUNT;
int foreground = 0;
char *pidfile = NULL;

static void newline_to_null(char *str)
{
	char *term = index(str, '\n');
	if (term)
		*term = '\0';
}

static void strip_domain(char *str)
{
	char *term = index(str, '.');
	if (term)
		*term = '\0';
}

static void build_node_desc(char *dest, size_t len,
		     const char *device, const char *hostname)
{
	char *end = dest + len-1;
	const char *field;
	char *src = ibd_nd_format;

	while (*src && (dest < end)) {
		if (*src != '%') {
			*dest++ = *src++;
		} else {
			src++;
			switch (*src) {
			case 'h':
				field = hostname;
				while (*field && (*field != '.') && (dest < end))
					*dest++ = *field++;
				break;
			case 'd':
				field = device;
				while (*field && (dest < end))
					*dest++ = *field++;
				break;
			}
			src++;
		}
	}
	*dest = 0;
}

static int update_node_desc(const char *device, const char *hostname, int force)
{
	int rc;
	char nd[128];
	char new_nd[64];
	char nd_file[PATH_MAX];
	FILE *f;

	snprintf(nd_file, sizeof(nd_file), SYS_INFINIBAND "/%s/node_desc",
			device);
	nd_file[sizeof(nd_file)-1] = '\0';

	f = fopen(nd_file, "r+");
	if (!f) {
		syslog(LOG_ERR, "Failed to open %s\n", nd_file);
		return -EIO;
	}

	if (!fgets(nd, sizeof(nd), f)) {
		syslog(LOG_ERR, "Failed to read %s\n", nd_file);
		rc = -EIO;
		goto error;
	}
	newline_to_null(nd);

	build_node_desc(new_nd, sizeof(new_nd), device, hostname);

	if (!force && strncmp(new_nd, nd, sizeof(new_nd)) == 0) {
		syslog(LOG_INFO, "%s: no change (%s)\n", device, new_nd);
	} else {
		syslog(LOG_INFO, "%s: change (%s) -> (%s)\n",
			device, nd, new_nd);
		rewind(f);
		fprintf(f, "%s", new_nd);
	}

	rc = 0;
error:
	fclose(f);
	return rc;
}

static int set_rdma_node_desc(const char *hostname, int force)
{
	DIR *class_dir;
	struct dirent *dent;

	class_dir = opendir(SYS_INFINIBAND);
	if (!class_dir) {
		syslog(LOG_INFO, "Failed to open " SYS_INFINIBAND);
		return -ENOSYS;
	}

	while ((dent = readdir(class_dir))) {
		int retry = set_retry_cnt;
		if (dent->d_name[0] == '.')
			continue;

		while (update_node_desc(dent->d_name, hostname, force) && retry > 0) {
			syslog(LOG_ERR, "retrying set Node Description on %s\n",
				dent->d_name);
			retry--;
		}
	}

	closedir(class_dir);
	return 0;
}

static int read_hostname(int fd, char *name, size_t len)
{
	int rc;
	memset(name, 0, len);
	if (read(fd, name, len-1) >= 0) {
		newline_to_null(name);
		strip_domain(name);
		rc = 0;
	} else {
		syslog(LOG_ERR, "Read %s Failed\n", SYS_HOSTNAME);
		rc = -EIO;
	}
	return rc;
}

static int process_opts(void *context, int ch, char *optarg)
{
	unsigned long tmp;
	switch (ch) {
	case 0:
		pidfile = optarg;
		break;
	case 'f':
		foreground = 1;
		break;
	case 't':
		tmp = strtoul(optarg, NULL, 0);
		if (tmp >= INT_MAX) {
			syslog(LOG_ERR,
				"Invalid retry rate specified: %lu s\n",
				tmp);
		} else {
			failure_retry_rate = (int)tmp;
		}
		break;
	case 'r':
		tmp = strtoul(optarg, NULL, 0);
		if (tmp >= INT_MAX) {
			syslog(LOG_ERR,
				"Invalid retry count specified: %lu\n",
				tmp);
		} else {
			set_retry_cnt = (int)tmp;
		}
		break;
	default:
		return -1;
	}
	return 0;
}

#if HAVE_UDEV_LOGGING
#define MSG_MAX 2048
static void udev_log_fn(struct udev *ud, int priority, const char *file, int line,
		const char *fn, const char *format, va_list args)
{
	int off = 0;
	char msg[MSG_MAX];
	off = snprintf(msg, MSG_MAX, "libudev: %s:%d %s",
			file, line, fn);
	if (off < MSG_MAX-1)
		vsnprintf(msg+off, MSG_MAX-off, format, args);
	syslog(LOG_ERR, "%s", msg);
}
#endif

static void setup_udev(void)
{
	udev = udev_new();
	if (!udev) {
		syslog(LOG_ERR, "udev_new failed\n");
		return;
	}

#if HAVE_UDEV_LOGGING
	udev_set_log_fn(udev, udev_log_fn);
	udev_set_log_priority(udev, LOG_INFO);
#endif
}

static int get_udev_fd(void)
{
	mon = udev_monitor_new_from_netlink(udev, "udev");
	if (!mon) {
		syslog(LOG_ERR, "udev monitoring failed\n");
		return -1;
	}

	udev_monitor_filter_add_match_subsystem_devtype(mon, "infiniband", NULL);
	udev_monitor_enable_receiving(mon);
	return udev_monitor_get_fd(mon);
}

static void process_udev_event(int ud_fd, const char *hostname)
{
	struct udev_device *dev;

	dev = udev_monitor_receive_device(mon);
	if (dev) {
		const char *device = udev_device_get_sysname(dev);
		const char *action = udev_device_get_action(dev);

		syslog(LOG_INFO, "Device event: %s, %s, %s\n",
			udev_device_get_subsystem(dev),
			device, action);

		if (device && action
		    && strncmp(action, "add", sizeof("add")) == 0)
			update_node_desc(device, hostname, 1);

		udev_device_unref(dev);
	}
}

static void monitor(void)
{
	char hostname[128];
	int hn_fd;
	int rc;
	struct pollfd fds[2];
	int numfds = 1;
	int ud_fd;

	ud_fd = get_udev_fd();
	if (ud_fd >= 0)
		numfds = 2;

	while (1) {
		hn_fd = open(SYS_HOSTNAME, O_RDONLY);
		if (hn_fd < 0) {
			syslog(LOG_ERR,
				"Open %s Failed: retry in %d seconds\n",
				SYS_HOSTNAME, failure_retry_rate);
			sleep(failure_retry_rate);
			continue;
		}

		fds[0].fd = hn_fd;
		fds[0].events = 0;
		fds[0].revents = 0;

		fds[1].fd = ud_fd;
		fds[1].events = POLLIN;
		fds[1].revents = 0;

		rc = poll(fds, numfds, -1);

		if (rc > 0) {
			if (read_hostname(hn_fd, hostname, sizeof(hostname)) != 0)
				hostname[0] = '\0';

			if (fds[0].revents != 0)
				syslog(LOG_ERR, "Hostname change: %s\n", hostname);

			if (fds[1].revents != 0)
				process_udev_event(ud_fd, hostname);

			rc = set_rdma_node_desc((const char *)hostname, 0);
		} else {
			syslog(LOG_ERR, "Poll %s Failed\n", SYS_HOSTNAME);
			rc = -EIO;
		}

		close(hn_fd);

		if (rc)
			sleep(failure_retry_rate);
	}
}

static void remove_pidfile(void)
{
        if (pidfile)
		unlink(pidfile);
}

static void write_pidfile(void)
{
	FILE *f;
	if (pidfile) {
		remove_pidfile();
		f = fopen(pidfile, "w");
		if (f) {
			fprintf(f, "%d\n", getpid());
			fclose(f);
		} else {
			syslog(LOG_ERR, "Failed to write pidfile : %s\n",
				pidfile);
			exit(errno);
		}
	}
}

int main(int argc, char *argv[])
{
	int fd;
	char hostname[128];

	openlog("rdma-ndd", LOG_PID | LOG_PERROR, LOG_DAEMON);

	const struct ibdiag_opt opts[] = {
		{"retry_timer", 't', 1, "<retry_timer>",
			"Length of time to sleep when system errors occur "
			"when attempting to poll and or read the hostname "
			"from the system.\n"},
		{"retry_count", 'r', 1, "<retry_count>",
			"Number of times to attempt to retry setting "
			"of the node description on failure\n"},
		{"foreground", 'f', 0, NULL, "run in the foreground instead of as a daemon\n"},
		{"pidfile", 0, 1, "<pidfile>", "specify a pid file (daemon mode only)\n"},
		{0}
	};

	ibdiag_process_opts(argc, argv, NULL, "CPDLGtsKyevd", opts,
			    process_opts, "", NULL);

	if (!ibd_nd_format)
		ibd_nd_format = DEFAULT_ND_FORMAT;

	if (!foreground) {
		closelog();
		openlog("rdma-ndd", LOG_PID, LOG_DAEMON);
		if (daemon(0, 0) != 0) {
			syslog(LOG_ERR, "Failed to daemonize\n");
			exit(errno);
		}
		write_pidfile();
	}

	setup_udev();

	syslog(LOG_INFO, "Node Descriptor format (%s)\n", ibd_nd_format);

	fd = open(SYS_HOSTNAME, O_RDONLY);
	if (read_hostname(fd, hostname, sizeof(hostname)) != 0)
		hostname[0] = '\0';
	set_rdma_node_desc((const char *)hostname, 1);
	close(fd);

	monitor();

	remove_pidfile();

	return 0;
}
